ksql-examples/src/main/java/io/confluent/ksql/datagen/DataGenProducer.java

```diff
@@ -152,7 +106,89 @@ public void populateTopic(
     producer.close();
   }
 
+
+  // For test purpose.
+  protected Pair<String, GenericRow> generateOneGenericRow(
+      final Generator generator,
+      final AvroData avroData,
+      final Schema avroSchema,
+      final org.apache.kafka.connect.data.Schema ksqlSchema,
+      final SessionManager sessionManager,
+      final String key) {
+
+    final Object generatedObject = generator.generate();
+
+    if (!(generatedObject instanceof GenericRecord)) {
+      throw new RuntimeException(String.format(
+          "Expected Avro Random Generator to return instance of GenericRecord, found %s instead",
+          generatedObject.getClass().getName()
+      ));
+    }
+    final GenericRecord randomAvroMessage = (GenericRecord) generatedObject;
+
+    final List<Object> genericRowValues = new ArrayList<>();
+
+    SimpleDateFormat timeformatter = null;
+
+    /**
+     * Populate the record entries
+     */
+    String sessionisationValue = null;
+    for (Schema.Field field : avroSchema.getFields()) {
+
+      final boolean isSession = field.schema().getProp("session") != null;
+      final boolean isSessionSiblingIntHash =
+          field.schema().getProp("session-sibling-int-hash") != null;
+      final String timeFormatFromLong = field.schema().getProp("format_as_time");
+
+      if (isSession) {
+        final String currentValue = (String) randomAvroMessage.get(field.name());
+        final String newCurrentValue = handleSessionisationOfValue(sessionManager, currentValue);
+        sessionisationValue = newCurrentValue;
+
+        genericRowValues.add(newCurrentValue);
+      } else if (isSessionSiblingIntHash && sessionisationValue != null) {
+
+        // super cheeky hack to link int-ids to session-values - if anything fails then we use
+        // the 'avro-gen' randomised version
+        handleSessionSiblingField(
+            randomAvroMessage,
+            genericRowValues,
+            sessionisationValue,
+            field
+        );
+
+      } else if (timeFormatFromLong != null) {
+        final Date date = new Date(System.currentTimeMillis());
+        if (timeFormatFromLong.equals("unix_long")) {
+          genericRowValues.add(date.getTime());
+        } else {
+          if (timeformatter == null) {
+            timeformatter = new SimpleDateFormat(timeFormatFromLong);
+          }
+          genericRowValues.add(timeformatter.format(date));
+        }
+      } else {
+        final Object value = randomAvroMessage.get(field.name());
```

Comment:
Apart from the refactoring and the additional of optional schemas, this is the only change needed to support struct, right?

Reply:
Yes, it wasn't that much change.